mtasts/cache.go

```diff
@@ -0,0 +1,174 @@
+package mtasts
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/emersion/maddy/log"
+)
+
+func downloadPolicy(domain string) (*Policy, error) {
+	// TODO: Consult OCSP/CRL to detect revoked certificates?
+
+	resp, err := http.Get("https://mta-sts." + domain + "/.well-known/mta-sts.txt")
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Policies fetched via HTTPS are only valid if the HTTP response code is
+	// 200 (OK).  HTTP 3xx redirects MUST NOT be followed.
+	if resp.StatusCode != 200 {
+		return nil, errors.New("mtasts: HTTP " + resp.Status)
+	}
+
+	if !strings.HasPrefix(resp.Header.Get("Content-Type"), "text/plain") {
+		return nil, errors.New("mtasts: unexpected content type")
+	}
+
+	return readPolicy(resp.Body)
+}
+
+// Cache structure implements transparent MTA-STS policy caching using FS
+// directory.
+type Cache struct {
+	Location string
+	Logger   *log.Logger
+}
+
+var ErrNoPolicy = errors.New("mtasts: no MTA-STS policy found")
+
+// Get reads policy from cache or tries to fetch it from Policy Host.
+func (c *Cache) Get(domain string) (*Policy, error) {
+	_, p, err := c.fetch(domain)
+	return p, err
+}
+
+func (c *Cache) store(domain, id string, fetchTime time.Time, p *Policy) error {
+	f, err := os.Create(filepath.Join(c.Location, domain))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return json.NewEncoder(f).Encode(map[string]interface{}{
+		"ID":        id,
+		"FetchTime": fetchTime,
+		"Policy":    p,
+	})
+}
+
+func (c *Cache) load(domain string) (id string, fetchTime time.Time, p *Policy, err error) {
+	f, err := os.Open(filepath.Join(c.Location, domain))
+	if err != nil {
+		return "", time.Time{}, nil, err
+	}
+	defer f.Close()
+
+	data := struct {
+		ID        string
+		FetchTime time.Time
+		Policy    *Policy
+	}{}
+	if err := json.NewDecoder(f).Decode(&data); err != nil {
+		return "", time.Time{}, nil, err
+	}
+	return data.ID, data.FetchTime, data.Policy, nil
+}
+
+func (c *Cache) RefreshCache() error {
```

Comment:
I'm wondering whether refreshing the cache like this is a good idea. An alternative would be to refresh only when we need the MTA-STS policy (ie. we're sending an e-mail) and our cache is out-of-date.

Reply:
https://tools.ietf.org/html/rfc8461#section-10.2
See the third paragraph.

>   Because this attack is also possible upon refresh of a cached policy,
   implementers SHOULD NOT wait until a cached policy has expired before
   checking for an update; if senders attempt to refresh the cache
   regularly (for example, by fetching the current live policy in a
   background task that runs daily or weekly, regardless of the state of
   the "_mta-sts" TXT record, and updating their cache's "max age"
   accordingly), an attacker would have to foil policy discovery
   consistently over the lifetime of a cached policy to prevent a
   successful refresh.

Reply:
Oh, that makes a lot of sense.

Reply:
Though I missed that part about TXT record, it might be important to actually ignore it while refreshing cache.